### PR TITLE
Non-interactive setup

### PR DESCRIPTION
**Pull request by kkrynitski**
First, I want to say thank you guys, for that cool and very useful plugin. But for my needs (I suppose not only for my) I need non-interactive setup, cause that plugin is a part of debian package. So let me introduse setup, with non-interactive feature. So the work is not fully done. The purpose of the PR is to show way how I see it to be done, and to confirm that not only I intrested in such feature. Thank you for your attension.

Examples:

1.) sudo ./setup.py -i
inteactive mode, with EC2 metadata as defaults if possible (region, hostname, IAM)
2.) sudo ./setup.py -i -r <region_name> -H <hostname> 
intercactive mode, with overrading region and hostname
3.) sudo ./setup.py -i -a <access key> -s <secret_key>
interactive mode, with IAM user
4.) sudo ./setup  -i -a <access key> -s <secret_key> -P <creds_path>
interactive mode, with IAM user and custom aws credentials path

### CODE REVIEW

File: src/setup.py [hunk 1]
```diff
@@ -305,20 +306,36 @@ def __init__(self, credentials_path=None, access_key=None, secret_key=None, regi
...
+            self._configure_proxy_server_name_non_interactive()
+            self._configure_proxy_server_port_non_interactive()
+        else:
+            self._configure_region()
+            self._configure_hostname()
+            self._configure_credentials()
+            self._configure_proxy_server_name()
+            self._configure_proxy_server_port()
+            self._configure_push_asg()
+            self._configure_push_constant()
```
Comment (yimuniao) [inline src/setup.py]: asg and constant also need this non_interactive, could you add them?

Comment (kkrynitski) [inline src/setup.py]: Nice to hear that!!! So, my goal is to do the same functionality as interactive setup.   self._configure_proxy_server_port() and  self._configure_proxy_server_name() already added, working on self._configure_push_asg() and self._configure_push_constant() now.

File: src/setup.py [hunk 2]
```diff
@@ -337,13 +354,22 @@ def _configure_push_constant(self):
             self.config.push_constant = True
             self.config.constant_dimension_value = self._get_constant_dimension_value()
 
+    def _configure_region_non_interactive(self):
+        if self.region:
+            self.config.region = self.region
+        else:
+            self._configure_region()
+
     def _configure_region(self):
```
Comment (bedge) [inline src/setup.py]:
```
      def _configure_region(self, region=region): 
```
then
```
if region:
    self.config.region = region
else:
   <interactive stuff>
```

File: src/setup.py [hunk 3]
```diff
@@ -305,20 +306,36 @@ def __init__(self, credentials_path=None, access_key=None, secret_key=None, regi
...
     def run(self):
-        self._configure_region()
-        self._configure_hostname()
-        self._configure_credentials()
-        self._configure_proxy_server_name()
-        self._configure_proxy_server_port()
-        self._configure_push_asg()
-        self._configure_push_constant()
-        self._configure_plugin_installation_method()
+        if self.non_interactive:
```
Comment (bedge) [inline src/setup.py]: Just a thought on the non-interactive mechanism.
You could use use default args with a default of None and avoid the need to wrap the original function.
ie:
```
 self._configure_region(region=args.region)
```

File: src/setup.py [hunk 4]
```diff
@@ -565,6 +683,91 @@ def main():
                                  "Creating backup of the original configuration")
     REPLACE_WHITELIST_CMD = CMD(COPY_CMD.format(source=RECOMMENDED_WHITELIST, target=DEFAULT_PLUGIN_CONFIGURATION_DIR), "Replacing whitelist configuration")
 
+    parser = argparse.ArgumentParser(
+        formatter_class=argparse.ArgumentDefaultsHelpFormatter,
+        description='Script for custom installation process for collectd AWS CloudWatch plugin'
+    )
+    parser.add_argument(
+        '-i', '--non_interactive', required=False,
```
Comment (yimuniao) [inline src/setup.py]: I think it is better to use "-I", from my first view, "-i" means "interactive"

Comment (bedge) [inline src/setup.py]: Agreed, "-I" (uppercase "i") or "-n" is more conventional for non-interactive.
The uppercase being considered as a negation.

**Comment by yimuniao** [comment]
Please confirm this contribution is under the terms of the MIT license.

File: src/setup.py (hunk 4 above)
Comment (kkrynitski) [inline src/setup.py]: Agree, will change it to 'I'

**Comment by kkrynitski** [comment]
'-d' argument is not working now, it not designed for change in the final configuration file "/opt/collectd-plugins/cloudwatch/config/plugin.conf cause it was not in original setup. 
It will be something like that:

```    def _debug(self):
        logger.info('-=**********DEBUG**********=-')
        logger.info('PUSH ASG: {}'.format(self.config.push_asg))
        logger.info('PUSH CONSTANT: {}'.format(self.config.push_constant))
        logger.info('CONSTANT DIMENSION VALUE: {}'.format(self.config.constant_dimension_value))
        logger.info('CREDENTIALS PATH: {}'.format(self.config.credentials_path))
        logger.info('SECRET KEY: {}'.format(self.config.secret_key))
        logger.info('ACCESS KEY: {}'.format(self.config.access_key))
        logger.info('HOST: {}'.format(self.config.host))
        logger.info('PROXY NAME: {}'.format(self.config.proxy_server_name))
        logger.info('PROXY PORT: {}'.format(self.config.proxy_server_port))
        logger.info('CREDENTIALS PATH: {}'.format(self.config.credentials_path))
        logger.info('METHOD ONLY ADD PLUGIN: {}'.format(self.config.only_add_plugin))
        logger.info('USE RECOMMENDED CONFIG: {}'.format(self.config.use_recommended_collectd_config))
        logger.info('REGION: {}'.format(self.config.region)) ```

**Comment by kkrynitski** [comment]
I confirm that this contribution is under the terms of the MIT license.

**Comment by kkrynitski** [comment]
@yimuniao About '-d', now I see that you mean ```# The debug parameter enables verbose logging of published metrics
debug = "False"```
Will check, how it drives from setup.py.

**Comment by kkrynitski** [comment]
Hi, so I've add some more functional and unittests (it's not cover everything, unfortunately have no time to work more on unittests), please check and accept PR if everything fine.
Thank you.

**Comment by yimuniao** [comment]
Yes, Thanks a lot for this patch. I am applying for merging.

**Comment by dgouldin** [comment]
I've resolved the merge conflicts and added non-interactive args for `enable_high_resolution_metrics` and `flush_interval_in_seconds` in my fork: https://github.com/dgouldin/collectd-cloudwatch/tree/non_interactive_setup

All tests pass, but I've not done any manual testing so caveat emptor.

My fork is MIT licensed.

**Comment by yimuniao** [comment]
Thanks for your response. But looks like there is still conflict. The "Merge pull request" button is disabled.

**Comment by yimuniao** [comment]
I create a new pull request from your repo, https://github.com/awslabs/collectd-cloudwatch/pull/52

Is that ok for you?

**Comment by dgouldin** [comment]
That's fine.

**Comment by yimuniao** [comment]
Hi dgouldin, 

Two comments for your codes.
please check https://github.com/awslabs/collectd-cloudwatch/pull/52
Please fix it. I want to merge the codes as soon as possible.

Thanks
Liang

**Pull request merged**